File: cpp/problems/leetcode/0096_unique_binary_search_trees/solution.cpp
// LeetCode 96
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int numTrees(int n) {
        vector<long long> g(n + 1, 0);
        g[0] = 1;
        for (int i = 1; i <= n; ++i)
            for (int j = 1; j <= i; ++j)
                g[i] += g[j - 1] * g[i - j];
        return (int)g[n];
    }
};

int main() {
    assert(Solution().numTrees(3) == 5);
    assert(Solution().numTrees(1) == 1);
    assert(Solution().numTrees(0) == 1);
    assert(Solution().numTrees(5) == 42);
    cout << "0096 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0096_unique_binary_search_trees/solution.py
"""LeetCode 96: Unique Binary Search Trees."""

from __future__ import annotations


class Solution:
    def numTrees(self, n: int) -> int:
        g = [0] * (n + 1)
        g[0] = 1
        for i in range(1, n + 1):
            for j in range(1, i + 1):
                g[i] += g[j - 1] * g[i - j]
        return g[n]


if __name__ == "__main__":
    assert Solution().numTrees(3) == 5
    assert Solution().numTrees(1) == 1
    assert Solution().numTrees(0) == 1
    assert Solution().numTrees(5) == 42
    print("0096 OK")
